src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala

```diff
@@ -79,18 +85,19 @@ case class SingleWorkflowRunnerActor(source: WorkflowSourceFiles,
       log.info(s"$tag: workflow ID UUID($id)")
       workflowManager ! SubscribeToWorkflow(id)
 
-      if (getAbortJobsOnTerminate) {
-        Runtime.getRuntime.addShutdownHook(new Thread() {
-          override def run(): Unit = {
-            workflowManager ! WorkflowAbort(id)
-            log.info(s"$tag: Waiting for workflow $id to abort...")
-            while(stateName != Done)
-              Thread.sleep(1000)
-            log.info(s"$tag: Workflow $id aborted.")
-          }
-        })
+      if (abortJobsOnTerminate) {
+        sys.addShutdownHook {
+          workflowManager ! WorkflowAbort(id)
+          log.info(s"$tag: Waiting for workflow $id to abort...")
+          Await.ready(donePromise.future, Duration.Inf)
+        }
       }
       stay using data.copy(id = Option(id))
+    case Event(Transition(_, _, WorkflowAborted), data) =>
+      // The .get is safe because the shutdown hook is only registered when the workflow ID
+      // has been set in the state data.
+      log.info(s"$tag: Workflow ${data.id.get} aborted.")
```

Comment:
With more state data types and without so many `Option`s :smile:  Haven't thought it through any more than that yet.